### PR TITLE
Paper: Fluctuation X-ray Scattering real-time app

### PR DESCRIPTION
**Pull request by AntoineDujardin**
The purpose of this PR is to submit a draft of our paper on our Fluctuation X-ray Scattering real-time app.

### CODE REVIEW

**Comment by stargaser** [comment]
Hello @AntoineDujardin.  None of the assigned reviewers have so far responded to our invitation to review so I will be stepping in to help review your submission. I attended a small part of NERSC training for new users earlier this week, as it happens, and I am gratified to see some of what I learned in your paper. 

The proceedings committee decided to give reviewers this weekend to allow a bit more time -- I'll complete my review by then.

**Comment by AntoineDujardin** [comment]
Thanks, @stargaser, for the update and stepping in.

**Comment by himelmallick** [comment]
Hello, I am not sure I can do an in-depth review as a non-expert in this specific domain of FXS data analysis, but I am happy to provide a few high-level comments.

First, I would like to apologize for the unintentional delay as a result of a technical glitch that made me miss out on the initial email that was sent to all the assigned reviewers. I am still glad that I am able to review the paper as the saying goes 'better late than never' :)

Coming back to the review, overall, the authors have done a great job in showcasing a compelling application of Python which is easy to follow even as a non-expert. My specific comments are as follows:

* Figure 2 is referenced as `Figures 2` which needs to be fixed.
* Figure 1 is taken from Wikipedia without referencing the source.
* The major time improvements that are summarized in the text across instances can be summarized in a table, which will be a helpful addition for the readers' quick glance.
* All the three figures require a short description of what is being shown (in addition to the captions).
* The Conclusion section should be expanded to include some limitations (if any) and immediate next steps and potential future directions.
* In page 3, there is a sentence that ends with three dots (`However, such an implementation does not sound trivial using NumPy. . .`) which should be fixed.
* Phrases like `We will` have appeared in multiple places throughout the text including the Abstract which should be changed to present tense to go with the overall flow of the narrative.
* Please provide a reference to the `Convolution Theorem`.
* Please include a brief summary of how the correlation function was filtered and reduced to reconstruct Figure 3.
* Finally, as mentioned in the text `the size of the correlation function only increases with the desired resolution and is independent from the number of images in the dataset`, an example application with varying number of images will be helpful to make this point more clear.

**Comment by deniederhut** [comment]
Hey @AntoineDujardin ! Thanks to some awesome behind the scenes work by @stargaser, we now have experimental support for publishing papers with ORCIDs. If you have one and would like to add it to your paper, it's as simple as adding an `:orcid:` tag under your name in the paper header. [You can see an example of how to do this in the `00_vanderwalt` example paper](https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2020/papers/00_vanderwalt/00_vanderwalt.rst). If you don't have one, no worries! You can still publish your paper as is.

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+   fluctuation scattering, free electron laser, real-time analysis
+
+Introduction
+------------
+
+LCLS-II, an upgrade to LCLS
++++++++++++++++++++++++++++
+
+The Linac Coherent Light Source (LCLS) at the SLAC National Accelerator Laboratory is an X-ray Free Electron Laser providing femtosecond pulses with an ultrabright beam approximately one billion times brighter than synchrotrons. Such a brightness allows to work with much smaller sample sizes while the shortness allows imaging below the rotational diffusion time of the molecules. With pulses of such an unprecedented brightness and shortness, scientists are able to take snapshots of single macromolecules without the need for crystallization at ambient temperature.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: Consider adding a general reference to the LCLS.

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+However, LCLS experiments are typically high-risk / high-reward and involve novel setups, varying levels of requirements, and last only for a few days of beam time. The novelty in the science can require adaptations in the algorithms, requiring the data analysis itself to be highly flexible. Furthermore, we want to give users as much freedom as possible in the way they analyze their data without expecting them to have a deep knowledge of large-scale computer programming.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: In the first sentence, to make the list to be all nouns, consider changing "last only for a few days" to "durations of only a few days".

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+Therefore, we require real time analysis, high performance computing capabilities and a complex pipeline, while at the same time requiring enough flexibility to adapt to novel experimental setups and analysis algorithms. We believe Python helps us in this compromise pretty well.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: Consider shortening "while at the same time" to "while" (a suggestion from M. Pacer's proselint package).

End of sentence, "pretty well" is quite informal. Consider "Python helps us handle the competing demands" or similar rewording.

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+FXS: an example analysis requiring HPC
+++++++++++++++++++++++++++++++++++++++
+
+While a variety of experiments that can be performed at LCLS, we will here focus on one specific example: Fluctuation X-ray Scattering (FXS).
+
+X-ray scattering of particles in a solution is a common technique in the study of the structure and dynamics of macromolecules in biologically-relevant conditions and gives an understanding of their function. However, traditional methods currently used at synchrotrons suffer from the fact that the exposure time is longer than the rotation time of the particle, leading to the capture of angularly-averaged patterns.
+FXS techniques fully utilize the femtosecond pulses to measure diffraction patterns from multiple identical macromolecules below the sample rotational diffusion times (Fig. :ref:`fig:fxs`). The patterns are then collected to reconstruct a 3D structure of the macromolecule or measure some of its properties. It has been described in the late 1970s :cite:`Kam1977,Kam1981` and has then been performed at LCLS before the upgrade :cite:`Pande2018,Kurta2017,Mendez2014,Mendez2016`.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: Consider omitting "the" from "fullly utilize the femtosecond pulses".

Last sentence, I feel unsure what the leading "It" means, perhaps "FXS was first described in the late 1970s".

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+In FXS, each diffraction pattern contains several identical particles in random orientations. Information about the structure of the individual particle can be recovered by studying the two-point angular correlation of the data. To do so, the 2D images are expanded in a 3D, orientation-invariant space, where they are aggregated using the following formula:
+
+.. math::
+   :label: eq:intro
+
+   C_2(q, q', \Delta\phi) = \frac{1}{2 \pi N} \sum_{j=1}^N \int_0^{2 \pi} I_j(q, \phi) I_j(q', \phi+\Delta\phi) d\phi
+
+where :math:`I_j(q, \phi)` represents the intensity of the j-th image, in polar coordinates. This correlator can then be used as a basis for the actual 3D reconstruction of the data (Fig. :ref:`fig:reconstruction`), using an algorithm described elsewhere :cite:`Donatelli2015,Pande2018`.
+
+Acceleration: getting the best out of numpy
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: Is NumPy generally capitalized on the N and the P in this sort of context?

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+-------------------------------------------
+
+The expansion/aggregation step presented in Equation (:ref:`eq:intro`) was originally the most computation intensive part of the application, representing the vast majority of the computation time. The original implementation was processing each :math:`I_j(q, \phi)` image one after the other and aggregating the results. This resulted in taking 424 milliseconds per image using numpy functions and slightly better performances using numba. As we will illustrate in this section, rewriting this critical step allowed us to gain a factor of 40 in its speed, without any other libraries or tools.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: The timings, here and throughout this section, do not include what platform these items are running on -- I find myself wishing to know. Also, what is the typical image size?

Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: Here is the first introduction in text of NumPy and Numba. Please cite the appropriately. For NumPy I found references on [this SciPy page](https://www.scipy.org/citing.html), and for Numba there is [this paper](https://dl.acm.org/doi/10.1145/2833157.2833162). Also, I believe that Numba is capitalized.

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+.. code-block:: python
+
+  A = np.fft.fft(images[i], axis=-1)
+  for j in range(N_RAD_BINS):
+      for k in range(N_RAD_BINS):
+          for l in range(N_PHI_BINS):
+              C2[j, k, l] += A[j, l] * A[k, l].conj()
+
+which takes 49.1 seconds per image, i.e. about 100 times slower, in accordance with the stereotype of Python being much slower than other languages.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]:
```suggestion
which takes 49.1 seconds per image, i.e. about 100 times slower than the naive implementation, in accordance with the stereotype of Python being much slower than other languages.
```

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+
+taking 1.06 seconds, i.e. 29% faster than the C-ordered case and 40.0 times faster than the naive implementation.
+We could note that, at that speed, the main computation gets close to the time required to perform the Fast Fourier Transform, which is, in our case at least, faster on C-ordered (107 ms) than F-ordered (230 ms) data. Removing the FFT computation would yield an even starker contrast (977 ms vs. 499 ms), but would neglect the cost of the re-alignment.
+
+In conclusion, implementing using numpy or numba naively gives significant improvement on computational speed compared to pure Python, but there is still a lot of room for improvement. On the other hand, such improvement does not necessarily require using fancier tools. In our case, we showed that batching our computation helped in the numba case. From there, a batched numpy expression looked interesting. However, it required optimizing the mathematical formulation of the problem to come up with a canonical expression, which could then be handed over to numpy. Last but not least, the memory layout can have a sizable impact on the computation, while being easy to tweak in numpy.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: "Last but not least" was flagged by the proselint tool and could be replaces with "Finally".

File: papers/antoine_dujardin/antoine_dujardin.rst [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+To enable the task-based system, it is necessary to separate the data handling from the data itself. This reification of the data flow is achieved by declaring *regions*, similar to multi-dimensional Pandas dataframes :cite:`McKinney2010`. Regions contain *fields*, each of which being similar to and exposed as a numpy array. Regions can be partitioned into subregions, which can be processed by different tasks, allowing the parallelism.
+
+We scale up to 64 nodes of NERSC’s Cori Haswell using Pygion, with 10 to 30 processes per node, to reach a throughput of more than 15,000 images per second, as illustrated in Figures :ref:`fig:scaling`. Compared to an equivalent MPI implementation, Pygion scales out of the box as it manages the load-balancing and provides high-level parallelization constructs. These constructs make it easy to rapidly explore different partitioning strategies, without writing or rewriting any communication code. This enabled us to quickly find a strategy that scales better than the straightforward but ultimately suboptimal strategy that we initially developed.
```
Comment (stargaser) [inline papers/antoine_dujardin/antoine_dujardin.rst]: "Figures" -> "Figure".

I believe "Burst Buffer" is an SSD drive. Looking at Figure 2, should this be faster than the Lustre filesystem? Some discussion in the text here about why these two cases are shown, would be very helpful.

**Review by stargaser — COMMENTED** [review]
Thanks to @himelmallick for many good review comments, which I endorse.

I am likewise not an expert in the subject matter. I enjoyed reading the paper, especially the section on accelerating the code, and the conclusion of optimizing the math of the problem rather than the code.

I've made some additional suggestions on citations and wording inline.

**Comment by deniederhut** [comment]
Hi @AntoineDujardin ! Have you had a chance to incorporate your reviewers' feedback?

**Comment by AntoineDujardin** [comment]
Hi All! Thanks for the feedback, review, and other info. I am getting to it now, as I wasn't available recently.

**Comment by AntoineDujardin** [comment]
Hi @himelmallick,
>     * Figure 1 is taken from Wikipedia without referencing the source.
The Wikipedia figure happens to be from one of the authors. How should re reference that?

**Comment by himelmallick** [comment]
> The Wikipedia figure happens to be from one of the authors. How should re reference that?

I think referring the original article would suffice if you have that information.

File: papers/antoine_dujardin/antoine_dujardin.rst (hunk 4 above)
Comment (AntoineDujardin) [reply papers/antoine_dujardin/antoine_dujardin.rst]: We wanted to stress on the fact that we fully use the fact that these pulses are "femtosecond", contrarily to other experimental setups. I feel like removing the "the" removes that stress. Do you have another suggestion to make that point clearer?

File: papers/antoine_dujardin/antoine_dujardin.rst (hunk 6 above)
Comment (AntoineDujardin) [reply papers/antoine_dujardin/antoine_dujardin.rst]: I'm adding the fact that the tests were run on Cori Haswell.
Regarding the image size, the parameters are given after the first code snippet. We assume here that the images are already binned into polar coordinates in the preprocessing.

**Comment by AntoineDujardin** [comment]
We are still working on a discussion on the post-processing leading to the reconstruction and on the future works.

**Comment by deniederhut** [comment]
Hi everyone! Friendly reminder that the open review period ends 2 July -- the committee will be making final decisions on each paper this Friday.

**Comment by AntoineDujardin** [comment]
I think the comments have now been addressed.

**Comment by deniederhut** [comment]
Hi @stargaser and @himelmallick ! Are you happy with the recent updates to the paper?

File: papers/antoine_dujardin/antoine_dujardin.rst (hunk 4 above)
Comment (stargaser) [reply papers/antoine_dujardin/antoine_dujardin.rst]: Apologies that I didn't see this question before -- it's a minor point, I'm fine with leaving "the" in there after thinking about it.

**Comment by stargaser** [comment]
Many thanks @AntoineDujardin for the revisions. I am happy with the paper -- looks good to me!

**Comment by himelmallick** [comment]
Likewise, I am happy with the edits and changes!

**Comment by AntoineDujardin** [comment]
Thanks for the reviews!
Two technical questions for @stargaser and @deniederhut:

- Regarding the Video Link, I left it as "ToBeDetermined". Can I assume that it'll be corrected at publication or should I remove it completely?

- Also, as I mentioned in issue #414, we'd ideally have wanted more than one corresponding author (me + C.H. Yoon), but the framework only seem to allow 1. Furthermore, if I only have an email field on the last author (C.H. Yoon only), the asterisk is still attached to the first author's name regardless. Is that something that can be easily fixed?

**Comment by stargaser** [comment]
For the video link, my understanding is that one of the conference chairs will insert the YouTube link for the video for this talk when that is available. I believe it is fine to leave it in.

For the corresponding author issue, I am not confident that can be easily fixed. I'll take a look at it.

**Comment by stargaser** [comment]
@AntoineDujardin I believe if you put a line with `:corresponding:` after the entry for your email, and again for your coauthor C.H. Yoon, the footnote will list both of you as corresponding authors. Right now there's not an entry and in that case the first author is assumed to be the sole corresponding author. 

I tried it on the sample `00_vanderwalt` paper and it listed two email addresses. Would you please try it on this paper?

**Comment by AntoineDujardin** [comment]
It's a bit hacky, but I got something that works for us. Thanks a lot!

**Comment by AntoineDujardin** [comment]
@deniederhut, I thought the video link would be edited once available but I see that it hasn't been... would it be possible to fix that?

**Pull request merged**